#55 Refactor of config

microwavekonijn opened the pull request:
Original PR #48 
Issues #55 #51 #50

microwavekonijn committed:
Refactored configuration

This refactor implements new config system and resolved some problem related to that. It also removed obsolete configurations and added some that are missing. Environment variables can now be used to overwrite any configuration if neccesary.

microwavekonijn committed:
Fixed test

Created some code to fake the config module in such a way that any configuration can overwritten and be reset easily.

Also set NODE_ENV for the docker containers(production, staging, and testing).

microwavekonijn committed:
Moved subbot status

The booted and busy statusses for subbots are moved to a more appropriate place.

microwavekonijn committed:
Merge branch 'develop' into improve-config

# Conflicts:
#	.gitignore
#	config/config.js
#	config/development/channels.js
#	config/development/communityGames.js
#	config/development/features.js
#	config/development/generalConfig.js
#	config/development/recreationalGames.js
#	config/production/features.js
#	config/production/generalConfig.js
#	config/staging/features.js
#	config/staging/generalConfig.js
#	docker-compose-test.yml
#	package.json
#	test/mocha.opts

microwavekonijn committed:
Resolved merge conflicts

- Added environment variables to appropriate compose files
- Removed token fix for subbots
- Added Faker to be included during the tests

microwavekonijn committed:
Temporary fix env vars

- Sets the DISCORD_API_TOKEN & SUBBOTS env variables, thought the should be set directly in travis CI

microwavekonijn committed:
Fix for dumb mistake :)

microwavekonijn committed:
Another dumb fix :\

microwavekonijn committed:
Tempory fix did not work :(

The env variables need to be set directly in Travis

JamesLongman commented on .travis.yml:
```diff
@@ -4,8 +4,6 @@ services:
   - docker
 
 before_script:
-    - bash ./docker/CISetup.sh $token $subBotID1 $subBotID2 $subBotID3 $subBotID4 $subBotID5 $subBotID6 $subBotToken1 $subBotToken2 $subBotToken3 $subBotToken4 $subBotToken5 $subBotToken6
```
Builds on travis aren't getting tokens because travis needs to set them here

microwavekonijn replied:
I changed that, now the docker compose file passes the DISCORD_API_TOKEN and SUBBOTS directly to the container and the new config system will overwrite the token and subBots config. My fix was to use a bash script to do set the two new vars, but they don't allow me to set global ones.

I can change the compose file, but it is easier and more logical to set them correct in travis.

JamesLongman commented on docker/CISetup.sh:
```diff
@@ -1,40 +0,0 @@
-# Copyright © 2018 DIG Development team. All rights reserved.
-
-# Hardcoded location of envConfig.json in travis build
-configFile="/home/travis/build/JamesLongman/DIGBot/config/envConfig.json"
```
I don't think this file should be deleted, but this target should be changed

JamesLongman commented on docker/CISetup.sh:
```diff
@@ -1,40 +0,0 @@
-# and subsequently overrite 13 versions of the config file. There is most likely a much
-# more efficient way of altering the config arguements to cut down on build time
-# Issue #50, build time for these lines: ~26 seconds
-jq -c '.token = $newVal' --arg newVal $token $configFile > tmp.$$.json && mv tmp.$$.json $configFile
```
the ."token"/".subBots.subBot1.id" sections here may need to be changed also to match the new config

microwavekonijn committed:
Added testing config

microwavekonijn committed:
Fixes subbot test

- Removed finally as it is not supported in node 7.6

microwavekonijn committed:
Corrected check if subbots are disabled

Now subbots are disabled when there are no subbots in the config instead of false

microwavekonijn committed:
Fixed testing

- The version should be read directly from the package.json file
- Disabled channel enforcement and mcs(didn’t carry over from the CI PR)
- Subbots now check correctly if there are any subbots configured
- Test which checks if the subbot feature is disabled turned off as the new config system breaks it
- Swapped the checks in passBot and moved currentBots++ to more appropriate location

microwavekonijn committed:
Belongs to prev commit

JamesLongman committed:
expanded on subBot tests to illuminate travis issue source

JamesLongman committed:
Temporary console logging to show what travis is actually recieving as the SUBBOTS env variable

microwavekonijn committed:
I can’t rely on my past self :|

- Fixed environment passed by docker compose (was SUBBOTS, corrected to SUBOTS_CONFIG)
- Added SUBBOTS_LIMIT env var which can overwrite subBotLimit config

microwavekonijn committed:
Removed Logging

JamesLongman commented on config/config-faker.js:
```diff
@@ -0,0 +1,46 @@
+const _ = require("lodash");
```
As far as I can tell this file is purely used for mocking the config in tests and is not used in any live source code. If this is the case can we move it to test/mocks/ or something.

microwavekonijn replied:
That's correct. I initially put it here as the file should not be tested and I have no clue if mocha will just skip over the file or still tries to test it.

JamesLongman replied:
I think mocha only runs code in describe and it blocks. It should be fine

JamesLongman commented on config/default.json:
```diff
@@ -0,0 +1,111 @@
+  "botUserID"               : "",
+  "subBotLimit"             : 5,
+  "subBots"                 : {
+    //    "bot2": {
```
This comment style is triggering my IDE and github and will probably trigger the style CI too. Can we use this style instead please:

```
"bot2": "Your first sub bot object",
"bot2": {
    
}
```

JSON doesn't officially support comments but will always overwrite the first time a key is used for the second so it can be used for commenting

I've also seen people do
```
"comment": "Your first sub bot object",
"bot2": {
}
```
But I think that once it's parsed once there probably won't be much performance impact, you're going to be rewriting to "comment" a lot anyway and that just adds an unnecesary property

microwavekonijn replied:
As far as I can tell the new config system doesn't have a problem with it as it uses a dependency to read json files. I will remove it from the default.json file but leave it in local.json.example without comments as a template.

microwavekonijn replied:
We should probably have a document with all the config keys, their format, and a description of what they do. Maybe issue an issue.

JamesLongman replied:
Yeah our docs in general are not good enough and probably substantially outdated, removing all json comments and issuing a doc improvement would work

JamesLongman commented on config/default.json:
```diff
@@ -0,0 +1,111 @@
+{
```
So if I understand correctly default.json is holding the entire config, and we're rewriting specific version controlled properties such as channel IDs using separate files version-controlled files, while keeping things like feature switches in default and rewriting env variables with local.json?

microwavekonijn replied:
So it will try to read environment variables first defined in custom-environment-variables.json, then it will fallback to local.json if no value is found, then {environment}.json (so if NODE_ENV is set to staging it will use staging.json), then it will check default.json. If the config key is defined nowhere it will throw an error. Every config key should be defined in default.json so we have a clear view of all the configurations even when it is not assigned a value there.

JamesLongman replied:
Ah very nice sounds good

JamesLongman commented on config/local.json.example:
```diff
@@ -0,0 +1,83 @@
+
+  "botUserID"               : "",
+  "subBots"                 : {
+    //    "bot2": {
```
Comments again should match a JSON style

JamesLongman commented on .gitignore:
```diff
@@ -3,5 +3,4 @@ vagrant/.vagrant
 !docker/db/.gitkeep
-config/development/*
-config/envConfig.json
+/config/local.json
```
config/local.json

JamesLongman commented on config/production.json:
```diff
@@ -0,0 +1,346 @@
+    "PS2": {
+      "name"          : "Planetside 2",
+      "roleids"       : [
+        "278671831857102849", // PS2/Guest
```
Comments again, I don't think this would even parse.

I can see this section would be difficult to comment in JSON so I'd be open to removing them entirely. We can easily identify roles when we need to with the !roles command

JamesLongman commented on config/staging.json:
```diff
@@ -0,0 +1,127 @@
+    ],
+    "root"                   : "/var/web/dignityofwar.com/bot-staging",
+    "port"                   : 1338,
+    "memoryNotificationLimit": 150
```
Lowering the limit on staging for stricter testing? Any basis for this figure or just seeing how it goes?

microwavekonijn replied:
I just copied this from the old configuration. Let me know if what value it should be if you want to change it.

JamesLongman replied:
I'm not objecting to it, it probably should be lower, I just wondered why 150, I guess we'll never know

JamesLongman commented on config/testing.json:
```diff
@@ -0,0 +1,191 @@
+    "channelPositionsEnforcement": false,
+    "disableCommandSpam"         : false,
+    "disableMentionSpam"         : false,
+    "modularChannelSystem"       : false, // Issue #49
```
comment

JamesLongman commented on config/testing.json:
```diff
@@ -0,0 +1,191 @@
+    "staffRoleID"        : "486335247966011392",
+    "superMuteRoleID"    : "486325210505805856",
+    "leaderRoles"        : ["486325667001270272"],
+    "root"               : "F:/Coding/DIG-discord"
```
This is actually my personal root for our old repo.

 test/project-files.js and /test/src/assets/pictures/picture-assets.js  tests should fail without a correct root but don't. This we should issue, let's just open an issue and leave a comment pointing to it

microwavekonijn replied:
I checked and I think it can be removed. It is used in one place outside of testing and that is to call local sfx and there are no local assets.

microwavekonijn replied:
Maybe this is a solution?
https://hk.saowen.com/a/ce9f3446eb03d3efd9aabd9acb50dbe1e03a3ca0a8b4097f34d106794632d81f

JamesLongman commented on docker-compose-test.yml:
```diff
@@ -10,3 +10,5 @@ services:
         entrypoint: /app/test.sh
         tty: true
+        environment:
+        - NODE_ENV=testing
```
Does local.json supersede testing.json?

microwavekonijn replied:
Yes. Check my comment on config/default.json.

JamesLongman commented on package.json:
```diff
@@ -18,21 +18,24 @@
     "eslint": "^3.5.0",
-    "mocha": "5.2.0",
+    "mocha": "^5.2.0",
+    "mock-require": "^3.0.2",
```
10/10 dependency, we're going to be fixing versioning soon for stabilities sake #45, I'm indifferent to if this is done now or then for these new dependencies

JamesLongman commented on src/lib/commands/stats.js:
```diff
@@ -38,7 +39,7 @@ function statsCalculations(started, msg, message) {
 
         // DIGBot Version
-        let version = '**Version:** ' + config.getConfig().version;
+        let version = '**Version:** ' + pjson.version;
```
Good, let's change the version though to 0.1.0

JamesLongman commented on src/lib/sub-bots/sub-bots.js:
```diff
@@ -45,26 +46,25 @@ module.exports = {
                 reject('The maximum number of subBots are currently running');
                 return;
-            } else {
-                currentBots++;
```
I believe this is an error in asynchronous logic. The current code marks a subbot as busy in the instant that it is requested and one is available. This prevents two requests attempting to be fulfilled when a second request is made before the first is fulfilled when only one sub-bot remains.

Discord message -> API -> our code and login request -> login success response ≈ 1-2s

microwavekonijn replied:
I didn't move it to it's original location. So now it will only increase after it finds a subbot which is not busy, which makes more sense.

JamesLongman commented on src/lib/sub-bots/sub-bots.js:
```diff
@@ -76,6 +76,7 @@ module.exports = {
             let bot = new Discord.Client();
             bot.login(token)
                 .then(() => {
+                    currentBots++;
```
See comment above

JamesLongman commented on src/lib/sub-bots/sub-bots.js:
```diff
@@ -45,26 +46,25 @@ module.exports = {
+            }
+
+            // Probably best not to hit this, should be checked by what's calling it for better handling
+            if (!Object.keys(config.get('subBots')).length) {
```
(typeof config.get('subBots') !== "object" || !(Object.keys(config.get('subBots')).length > 0))

microwavekonijn replied:
We don't have to test if subBots is an object, because it will always will be(if for some reason it isn't it should just throw a generic error). Second part I don't know if that is necessary, length will always be positive and the only value that we want it to return true is 0.

JamesLongman replied:
We test for type to guard against throwing errors as much as we can throughout our code we always try to handle unexpected returns gracefully. You're right about the second part though because the first is there we can simplify it to:

(typeof config.get('subBots') !== "object" || Object.keys(config.get('subBots')).length == 0)

microwavekonijn replied:
I would be care-full with that. I agree that we want to catch errors and we should probably implement some system where we define different error types which are caught and resolved. However when something is wrong it should throw an error, for instance I found that the environment config key was misspelled and no-one would have noticed. The new system threw an error and it came to my attention.

microwavekonijn replied:
src/lib/discord/discordbot.js line 78. `if (config.getConfig().enviroment === 'development') {`

JamesLongman replied:
We have a system in logger.js, we throw errors when we need to but it's better if we're the ones throwing them so we can add additional info.

I agree it's not about sweeping these things under the rug though, it's properly handled with reject('The sub bot feature is disabled or there are no subBots on file'); Which is literally what this if statement is testing, it both logs it and very noticeably informs the user.

Handling errors gracefully will often allow easy testing of failure cases too where throwing errors makes testing more complex

JamesLongman replied:
Oooooh waaaaait I get it now, !0 is true. Yeah ```(!Object.keys(config.get('subBots')).length) ``` is fine, I would like to properly gaurd property calls so an acceptable compromise to me seeing at it is a local var would be to ```(!Object.keys(config.get('subBots')).length) ``` in pass bot and ready functions but before it runs for the first time in ready do ``` if (typeof config.get('subBots') !== "object") { logger.error(TAG, "subBots property not object") } ```

We're still throwing an error, but it's a more graceful exit than a crash as the bot's actually informing us and throwing additional info

microwavekonijn replied:
How far do we want to go with this? It will still throw an generic error when it is passed incorrectly using an env var. Also this is not the only case channels.positions.text/voice, recreationalGames, and communityGames should also be objects. I don’t really see a point to cover all possible senarios of incorrect usage, as it makes things just more complicated. Should we for instance also check that every subbot is an object and has an id, and token?

JamesLongman replied:
I personally want to go far enough to cover most common instances of errors, as I would much rather throw stack trace alerts on discord and fail gracefully than risk impacting database calls or having to ssh because the bot is constantly restarting and we don't know why. But given that this is an env variable and a less likely point of failure I'mma just gunna merge this rather than argue any more so we can get on with both this project and our lives :)

JamesLongman commented on src/lib/sub-bots/sub-bots.js:
```diff
@@ -89,18 +90,25 @@ module.exports = {
     ready: function() {
         crashHandler.logEvent(TAG, 'ready');
-        if (!config.getConfig().subBots) {
+        if (!Object.keys(config.get('subBots')).length) {
```
(typeof config.get('subBots') !== "object" || !(Object.keys(config.get('subBots')).length > 0))

JamesLongman commented on test/mocha.opts:
```diff
@@ -8,5 +8,8 @@
 # Retry failed tests 3 times (useful if an error is just because of a dodgy API connection)
 --retries 3
 
+#Creates a global variable Faker which allows for easy modification of the config during testing
```
Space:
```
# Creates
```

JamesLongman commented on test/src/lib/commands/mentions.js:
```diff
@@ -45,40 +45,36 @@ describe('commands/mentions.js', function() {
+        Faker.setFakeProperty('features.disableMentionSpam', true);
         mentions.execute(msg).should.be.false;
-        config.setProperty('features', original);
+        Faker.resetFake();
```
This is much more graceful

JamesLongman commented on test/src/lib/sub-bots/sub-bots.js:
```diff
@@ -60,48 +73,51 @@ describe('sub-bots/sub-bots.js', function() {
-            result.should.eql('The sub bot feature is disabled or there are no subBots on file');
-        });
-    });
+    // Test doesn't work as subbots object in the config is not the same object
```
Firstly, omg y u no /* comment */ that's like a solid minute of commenting

Secondly, we'll issue this, probably just need to add a feature switch or something

microwavekonijn replied:
Srry :| I just select a block of code and then press cmd + / (windows ctrl + /) and it will make all lines into comments (I am very lazy). :)

JamesLongman replied:
Ah I see, I was not aware of such shortcut magic

JamesLongman reviewed (changes requested):
Looks good no major reworking needed, just a few tweaks mostly style and some questions.

There are a couple of things we will need to issue but nothing major. This has broken my local build a bit but I think I just need to sort my config out as testing seems to be alright on CI, I'm happy to sort out anything on develop so we can get on with staging

microwavekonijn committed:
Changes based on the review

- Moved currentBots++ to more appropriate location(when a unused subbot is found)
- Changed version in package.json to 0.1.0
- Removed comments from all config json files
- Comments in local.json.example are now empty templates
- Removed features from testing.json as they are the same as in default.json
- Added a space and removed a slash :)

microwavekonijn committed:
Moved Faker to testing

microwavekonijn committed:
Belongs to prev commit derp

microwavekonijn committed:
Added documentation for config
